### ISSUE TITLE
[Zoe] Long-term splitting of Zoe

### ISSUE THREAD

**Issue by katelynsills** [Agoric/agoric-sdk#1011]
This is the long-term plan for splitting up Zoe to guarantee payout liveness for users of Zoe contracts.

The short-term plan is defined here: https://github.com/Agoric/agoric-sdk/issues/553

There will be one "Zoe system" on our testnet. A single "Zoe system" is made up of three parts:

1. zoe-service
    - one per Zoe System
    - holds Zoe Invite Issuer (one per Zoe system)
    - coordination point
    - this is what the users will call `offer` on with their invite and their proposal and payments
    - has its own vat
2. zoe-contract-facet
   - a vat per contract instance
   - synchronously accessible from the contract code
3. zoe-vault
   - one vat per contract instance, meaning one per zoe-contract-facet as well
   - because math objects will grow over time and some will have a bug, and this shouldn't affect other contracts
   - holds the assets

Remains to be seen where the offer safety checks will be - possibly in both the zoe-contract-facet and the zoe-vault.

**Comment by erights** [comment]
An observation since our conversation:

If the vault does all[*] the enforcement needed to reason about offer safety and payout liveness, then it is the thing we need to verify, and therefore the thing we need to keep as simple as possible. But it is no longer the thing talking to the contract. The vault talks to the zoe-contract-facet (zcf) which talks to the contract. The relieves the pressure to keep zcf minimal and move functionality into helpers. More helpful non-fundamental conveniences could move into zcf without penalizing the verifiability of the vault.

In some sense, zcf is the all-encompassing "helper" for presenting the functionality of the vault conveniently to the contract.

[*] Offers come in through the zoe-service, to mostly be passed through to the vault. Offer safety and payout liveness depends on the correctness of this mostly-pass-through code as well, which we should also strive to keep simple and to verify.

zcf has no similar constraint because offer safety and payout liveness provides guarantees only to those who place offers. It provides no guarantees to the contract itself.

**Comment by katelynsills** [comment]
> zcf has no similar constraint because offer safety and payout liveness provides guarantees only to those who place offers. It provides no guarantees to the contract itself.

To be nitpicky, I think zcf does have a responsibility to the contract: zcf should fail-fast for reallocations that aren't offer safe or don't conserve rights, because any such allocations will fail in the vault if the zcf doesn't catch it.

**Comment by erights** [comment]
I agree. My point is that offer safety and payout liveness would not depend on zcf. Therefore, tardiness in verifying zcf, though serious, does not impede us from providing verified offer safety and payout liveness.

**Comment by katelynsills** [comment]
As of 2/8/2021, we have 1:zoe-service and 2:zoe-contract-facet, but we do not have 3:zoe-vault

What was our reasoning for wanting to have a zoe-vault per contract instance?

One of the reasons stated above in this issue was:
`because math objects will grow over time and some will have a bug, and this shouldn't affect other contracts`

If we are redesigning amountMath such that there are no longer amountMath objects that close over a brand, but instead, the user always imports the amountMath library, does this reason still remain?

**Comment by erights** [comment]
Yes. The issue is at the math helper level and independent of whether it is presented in amountMath instances or static methods. If, say, we admit a divisible space math helper because we vetted it wrongly, then contracts that never encounter that math helper shouldn't suffer from its misbehavior.

**Comment by katelynsills** [comment]
Ok, so this only seems necessary if we believe that some mathHelpers are more likely to flawed than others, and that we wish to protect the users of other mathHelpers against these flawed ones. This issue is already not in Beta, but if we are doing a thorough review of all the mathHelpers a user is able to use, I'm not sure it makes sense to protect against bad mathHelpers especially, compared to bad code in Zoe or ERTP proper.

**Comment by erights** [comment]
I think it is fine not to do this for beta. We should discuss but my first reaction is it is also fine not to do this for mainnet 1. But it is something we will want to do eventually, for at least this reason.
